### ISSUE TITLE
bsdthread_register error with (sigh) Websense/Forcepoint DLP client

### ISSUE THREAD

**Issue by bkrodgers** [hashicorp/vault#1593]
Obviously this is one you may be inclined to punt on.  However, I still wanted to open an issue to see what we can do.  Our company is sadly about to force Mac users to install the Websense/Forcepoint TRITON DLP client on our machines.   I'm testing it for them now.  As soon as it installs, vault stops working:

```
$ vault
fatal error: runtime: bsdthread_register error

runtime stack:
runtime.throw(0x150b500, 0x21)
	/usr/local/Cellar/go/1.6.2/libexec/src/runtime/panic.go:547 +0x90 fp=0x7fff5fbff9a0 sp=0x7fff5fbff988
runtime.goenvs()
	/usr/local/Cellar/go/1.6.2/libexec/src/runtime/os1_darwin.go:76 +0x8d fp=0x7fff5fbff9c8 sp=0x7fff5fbff9a0
runtime.schedinit()
	/usr/local/Cellar/go/1.6.2/libexec/src/runtime/proc.go:440 +0xa7 fp=0x7fff5fbffa10 sp=0x7fff5fbff9c8
runtime.rt0_go(0x7fff5fbffa40, 0x1, 0x7fff5fbffa40, 0x0, 0x1, 0x7fff5fbffb70, 0x0, 0x7fff5fbffb76, 0x7fff5fbffb92, 0x7fff5fbffba6, ...)
	/usr/local/Cellar/go/1.6.2/libexec/src/runtime/asm_amd64.s:138 +0x132 fp=0x7fff5fbffa18 sp=0x7fff5fbffa10
```

And as soon as I uninstall it, it starts working again.

It sounds like this is a known issue with some Go based programs and DLP tools such as this one.  There's a 2 year old discussion here:
https://groups.google.com/forum/#!topic/golang-nuts/HtcZc3wO-rw

And an issue here:
https://github.com/golang/go/issues/8801

As far as using the precompiled Vault binary, the workaround of unsetting DYLD_INSERT_LIBRARIES doesn't help me (I didn't have it set to begin with).

What is notable however, is that other Go based tools, such as docker and vagrant, are still working.  So I'm not sure if that might mean there is something that can be done at compile time to make Vault play well with this...thing.  On the issue linked above, there's a link to another issue with someone saying they were able to fix it.  https://github.com/tonytheodore/mxe/commit/b57e1bc835830f89cbfd60712dbe59892229dfb0

Of course I know you can't really test anything on your end, since you don't have this Websense client.  I'm happy to try to see what I can do, such as trying the change linked above locally.  I opened this issue to see if you had any thoughts and to track any progress.

**Comment by jefferai** [comment]
@bkrodgers Can you give https://www.dropbox.com/s/tgczodth74l84b0/darwin_amd64.zip?dl=0 a try and see if it works for you? (Also, make sure it's not broken on a non-DLP-"enhanced" machine?)

**Comment by bkrodgers** [comment]
Thanks Jeff.  Unfortunately it seems to be the same error.  The paths come up differently in the trace for whatever reason, but same files and line numbers.

The brew installed 0.6.0 is on top, the one you gave me is the second one.  

```
bkrodg-mac2:vault-test bkrodg$ /usr/local/bin/vault
fatal error: runtime: bsdthread_register error

runtime stack:
runtime.throw(0x150b500, 0x21)
	/usr/local/Cellar/go/1.6.2/libexec/src/runtime/panic.go:547 +0x90 fp=0x7fff5fbff9b0 sp=0x7fff5fbff998
runtime.goenvs()
	/usr/local/Cellar/go/1.6.2/libexec/src/runtime/os1_darwin.go:76 +0x8d fp=0x7fff5fbff9d8 sp=0x7fff5fbff9b0
runtime.schedinit()
	/usr/local/Cellar/go/1.6.2/libexec/src/runtime/proc.go:440 +0xa7 fp=0x7fff5fbffa20 sp=0x7fff5fbff9d8
runtime.rt0_go(0x7fff5fbffa50, 0x1, 0x7fff5fbffa50, 0x0, 0x1, 0x7fff5fbffb80, 0x0, 0x7fff5fbffb95, 0x7fff5fbffbb1, 0x7fff5fbffbc5, ...)
	/usr/local/Cellar/go/1.6.2/libexec/src/runtime/asm_amd64.s:138 +0x132 fp=0x7fff5fbffa28 sp=0x7fff5fbffa20


bkrodg-mac2:vault-test bkrodg$ ./vault
fatal error: runtime: bsdthread_register error

runtime stack:
runtime.throw(0x1530d20, 0x21)
	/goroot/src/runtime/panic.go:547 +0x90 fp=0x7fff5fbff9d0 sp=0x7fff5fbff9b8
runtime.goenvs()
	/goroot/src/runtime/os1_darwin.go:76 +0x8d fp=0x7fff5fbff9f8 sp=0x7fff5fbff9d0
runtime.schedinit()
	/goroot/src/runtime/proc.go:440 +0xa7 fp=0x7fff5fbffa40 sp=0x7fff5fbff9f8
runtime.rt0_go(0x7fff5fbffa78, 0x1, 0x7fff5fbffa78, 0x0, 0x0, 0x1, 0x7fff5fbffb98, 0x0, 0x7fff5fbffba0, 0x7fff5fbffbbc, ...)
	/goroot/src/runtime/asm_amd64.s:138 +0x132 fp=0x7fff5fbffa48 sp=0x7fff5fbffa40
```

**Comment by jefferai** [comment]
@bkrodgers This makes me wonder if, in fact, it's Go (rather than Vault) that has to be built with that environment variable set (or both). This is what https://github.com/tonytheodore/mxe/commit/b57e1bc835830f89cbfd60712dbe59892229dfb0 was doing, I believe, and I semi-suspected that from the linked issue/ML posts.

It's not too hard to build Go locally and then build Vault against it (by ensuring that the right binaries are first in your PATH). I unfortunately don't have an OSX machine right now, so I do need to punt this over to you for the moment. Any chance you can build Go from source with that env var set (to an empty string), then build Vault from source against that Go, also with the same env var set?

**Comment by bkrodgers** [comment]
Sure, I'll give that a shot and report back.

**Comment by bkrodgers** [comment]
Assuming I did things right, which is an "if," it doesn't seem to have helped.

First, I wasn't able to build anything (Go or Vault) with the DLP client installed, even with that set explicitly to ''.  That is, I tried to build Go with the command `DYLD_INSERT_LIBRARIES= ./all.bash`

I uninstalled the DLP client (which thankfully I can do for now since I'm helping test).  Rebuilt go with `DYLD_INSERT_LIBRARIES= ./all.bash` and that worked.  Did:
 ```
export PATH=<path-to-the-golang-i-just-built>/bin:$PATH
export GOPATH=$HOME/work
export PATH=$PATH:$GOPATH/bin
git clone git@github.com:hashicorp/vault.git $GOPATH/src/github.com/hashicorp/vault
cd $GOPATH/src/github.com/hashicorp/vault
DYLD_INSERT_LIBRARIES='' make bootstrap
DYLD_INSERT_LIBRARIES='' make
DYLD_INSERT_LIBRARIES='' make dev
```

The tests did on the plain "make" due to issues with postgres and mysql docker containers.  I'm running the native Mac docker beta, so perhaps that's why.  I was able to do `make dev` though and get a binary.  I ran it without DLP and it worked.  I reinstalled DLP, and it immediately failed again with the same error.

Does it sound like I built everything correctly?  Is putting my build of Go first in the path the only thing I needed to do to build Vault against it?  There was another Go install later in the path still, but I assume it would pick the custom one first.  I'm assuming this is a positive sign that I built it right, since the stack trace refers to the path where my Go source was:

```
bin/vault 
fatal error: runtime: bsdthread_register error

runtime stack:
runtime.throw(0x1530d40, 0x21)
	/Users/bkrodg/Documents/3ga/golang/src/runtime/panic.go:547 +0x90 fp=0x7fff5fbff960 sp=0x7fff5fbff948
runtime.goenvs()
	/Users/bkrodg/Documents/3ga/golang/src/runtime/os1_darwin.go:76 +0x8d fp=0x7fff5fbff988 sp=0x7fff5fbff960
runtime.schedinit()
	/Users/bkrodg/Documents/3ga/golang/src/runtime/proc.go:440 +0xa7 fp=0x7fff5fbff9d0 sp=0x7fff5fbff988
runtime.rt0_go(0x7fff5fbffa00, 0x1, 0x7fff5fbffa00, 0x0, 0x1, 0x7fff5fbffb30, 0x0, 0x7fff5fbffb3a, 0x7fff5fbffb56, 0x7fff5fbffb6a, ...)
	/Users/bkrodg/Documents/3ga/golang/src/runtime/asm_amd64.s:138 +0x132 fp=0x7fff5fbff9d8 sp=0x7fff5fbff9d0
```

**Comment by bkrodgers** [comment]
You can ignore the part about the tests failing due to docker issues.  I went to dockertest's issues repo and found this:  https://github.com/ory-am/dockertest/issues/51

Removing docker-machine from my Mac fixed that problem (just the test issue, none of the DLP issues).

**Comment by jefferai** [comment]
Hi @bkrodgers 

I do believe that your test steps above are correct. Which is sad, in this case.

At this point I have only a few thoughts/suggestions:

1) You could try contacting upstream Go and finding out if, in fact, there is a more correct/better way to do this. There was a post about linker flags (which others said didn't work) and the env var seemed like the most recent official way to fix it, but maybe not. A lot of developers are usually not amenable to having to deal with this kind of corporate strongarming but pleading/begging/puppy dog eyes can sometimes help.

2) You could see if there is a way to specify a whitelist for DLP -- a list of binaries where it does not preload itself. Often there are because vendors *know* that this kind of thing causes all sorts of problems, so it may be something that has to be administratively configured, but usually it's there.

3) You could run a VM, and run the Vault binary there. Not as convenient for sure, but it may just be the only way around.

If I think of anything else I'll let you know...

**Comment by jefferai** [comment]
Randomly flailing about with Google searches I found http://www.websense.com/support/article/kbarticle/Java-applications-no-longer-work-with-the-Websense-Web-Endpoint-since-a-Java-update which suggests that option (2) above does actually exist.

**Comment by bkrodgers** [comment]
1) "A lot of developers are usually not amenable to having to deal with this kind of corporate strongarming."  I'm not all that amenable to having to deal with it myself.  :-)  While we don't do a lot of Go development here, I'm equally concerned with the impact this will have on those who do need to do it.  It'd be nice to get the upstream Go toolchain fixed as well.  

2) Yeah, that is an option I'll be exploring with our folks.  I'm not quite sure how that works -- if it just goes off of the file name (in which case it furthers my case that this whole thing is dumb to have on developer workstations if it's that easy to bypass), or if it goes off of some kind of binary signature that would need to be continually update.  We shall see.

3) And yeah...that's an option too, which I've confirmed works.  Probably both because it's linux and this crash issue seems to be BSD related, but also probably because the methods the Websense client uses to inject itself into processes don't cross over into a VM.  Which again raises the question of why bother with dev machines.  I'd be interested to confirm that by running a Mac guest OS on vbox...need to figure out how to do that.

I'll keep seeing if I can find a way to compile vault so it works with WS installed.  It does seem that other Go programs have managed to fix this.  I'm looking through some other ones that I'm seeing working with WS installed to see if I can find any issues/commits they made to fix it.  I found issues for both Cloud Foundry's CLI and for Vagrant, but in both cases the issues didn't link to a commit that fixed it.

**Comment by jefferai** [comment]
It's odd that you found a fix for Vagrant, considering that it's not written in Go  :-)  But maybe some other non-language-specific workaround.

As for whether it makes sense to have this at all -- often times this is about checking a compliance box. You have to prove that you're running something. Then it's the vendor's fault if it doesn't work properly...e.g. if you whitelist an application and it doesn't check the signature, you can still claim that it was a failing in the security product, but that you adhered to security standards by having it installed.

There's a Websense endpoint client for Linux too, so it's not like they couldn't install it into Linux. But...I'll be quiet about it.  :-D

**Comment by bkrodgers** [comment]
Ahh, my bad on Vagrant...I thought all the hashicorp stuff was in Go.   :)  I hadn't looked at the code at all...I see now that it's Ruby.  

That being said though, there were two issues filed against Vagrant, and Mitchell said "This should be fixed in the next release, this is a Go issue."  
https://github.com/mitchellh/vagrant/issues/6914
https://github.com/mitchellh/vagrant/issues/6293

**Comment by jefferai** [comment]
Hah, I honestly have *no* idea. The GitHub repo doesn't say that it has any Go in it. My guess though is that by "fixed in the next release" what was meant was "we think the next Go release, which our next release will be built against, will fix it".

**Comment by bkrodgers** [comment]
I think I may have found a solution!  I tried this in `build.sh` on the gox command:

```
-ldflags "-linkmode=external -X github.com/hashicorp/vault/version.GitCommit='${GIT_COMMIT}${GIT_DIRTY}'" \
```

That built a version of Vault that worked with the Websense client installed!  And that both was without using the version of Go I'd built with the DYLD_INSERT_LIBRARIES='' prefix or setting that prefix on the Vault build.

I wanted to make sure it still cross-compiled properly, so I tried building with that change using `make bin`.  It looks like the external linkmode causes some problems with cross-compilation.  I'll look into this more tomorrow, but any ideas you've got are welcome.

```
13 errors occurred:
--> netbsd/arm error: exit status 2
Stderr: # github.com/hashicorp/vault
2016/07/07 19:39:58 cannot use -linkmode=external with -H netbsd

--> windows/amd64 error: exit status 2
Stderr: # github.com/hashicorp/vault
warning: unable to find runtime/cgo.a
/usr/local/Cellar/go/1.6.2/libexec/pkg/tool/darwin_amd64/link: running clang failed: exit status 1
ld: library not found for -ladvapi32
clang: error: linker command failed with exit code 1 (use -v to see invocation)


--> linux/arm error: exit status 2
Stderr: # github.com/hashicorp/vault
warning: unable to find runtime/cgo.a
/usr/local/Cellar/go/1.6.2/libexec/pkg/tool/darwin_amd64/link: running clang failed: exit status 1
ld: warning: ignoring file /var/folders/2f/t653xfss3x34b__mf65rtrtr9v8rmg/T/go-link-575194323/go.o, file was built for unsupported file format ( 0x7F 0x45 0x4C 0x46 0x01 0x01 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 ) which is not the architecture being linked (x86_64): /var/folders/2f/t653xfss3x34b__mf65rtrtr9v8rmg/T/go-link-575194323/go.o
Undefined symbols for architecture x86_64:
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)


--> freebsd/386 error: exit status 2
Stderr: # github.com/hashicorp/vault
warning: unable to find runtime/cgo.a
/usr/local/Cellar/go/1.6.2/libexec/pkg/tool/darwin_amd64/link: running clang failed: exit status 1
ld: warning: ignoring file /var/folders/2f/t653xfss3x34b__mf65rtrtr9v8rmg/T/go-link-381869002/go.o, file was built for unsupported file format ( 0x7F 0x45 0x4C 0x46 0x01 0x01 0x01 0x09 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 ) which is not the architecture being linked (i386): /var/folders/2f/t653xfss3x34b__mf65rtrtr9v8rmg/T/go-link-381869002/go.o
Undefined symbols for architecture i386:
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture i386
clang: error: linker command failed with exit code 1 (use -v to see invocation)


--> freebsd/arm error: exit status 2
Stderr: # github.com/hashicorp/vault
warning: unable to find runtime/cgo.a
/usr/local/Cellar/go/1.6.2/libexec/pkg/tool/darwin_amd64/link: running clang failed: exit status 1
ld: warning: ignoring file /var/folders/2f/t653xfss3x34b__mf65rtrtr9v8rmg/T/go-link-676707042/go.o, file was built for unsupported file format ( 0x7F 0x45 0x4C 0x46 0x01 0x01 0x01 0x09 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 ) which is not the architecture being linked (x86_64): /var/folders/2f/t653xfss3x34b__mf65rtrtr9v8rmg/T/go-link-676707042/go.o
Undefined symbols for architecture x86_64:
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)


--> freebsd/amd64 error: exit status 2
Stderr: # github.com/hashicorp/vault
warning: unable to find runtime/cgo.a
/usr/local/Cellar/go/1.6.2/libexec/pkg/tool/darwin_amd64/link: running clang failed: exit status 1
ld: warning: ignoring file /var/folders/2f/t653xfss3x34b__mf65rtrtr9v8rmg/T/go-link-627199716/go.o, file was built for unsupported file format ( 0x7F 0x45 0x4C 0x46 0x02 0x01 0x01 0x09 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 ) which is not the architecture being linked (x86_64): /var/folders/2f/t653xfss3x34b__mf65rtrtr9v8rmg/T/go-link-627199716/go.o
Undefined symbols for architecture x86_64:
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)


--> linux/386 error: exit status 2
Stderr: # github.com/hashicorp/vault
warning: unable to find runtime/cgo.a
/usr/local/Cellar/go/1.6.2/libexec/pkg/tool/darwin_amd64/link: running clang failed: exit status 1
ld: warning: ignoring file /var/folders/2f/t653xfss3x34b__mf65rtrtr9v8rmg/T/go-link-565567726/go.o, file was built for unsupported file format ( 0x7F 0x45 0x4C 0x46 0x01 0x01 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 ) which is not the architecture being linked (i386): /var/folders/2f/t653xfss3x34b__mf65rtrtr9v8rmg/T/go-link-565567726/go.o
Undefined symbols for architecture i386:
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture i386
clang: error: linker command failed with exit code 1 (use -v to see invocation)


--> openbsd/386 error: exit status 2
Stderr: # github.com/hashicorp/vault
warning: unable to find runtime/cgo.a
/usr/local/Cellar/go/1.6.2/libexec/pkg/tool/darwin_amd64/link: running clang failed: exit status 1
ld: unknown option: -nopie
clang: error: linker command failed with exit code 1 (use -v to see invocation)


--> windows/386 error: exit status 2
Stderr: # github.com/hashicorp/vault
warning: unable to find runtime/cgo.a
/usr/local/Cellar/go/1.6.2/libexec/pkg/tool/darwin_amd64/link: running clang failed: exit status 1
ld: library not found for -ladvapi32
clang: error: linker command failed with exit code 1 (use -v to see invocation)


--> netbsd/amd64 error: exit status 2
Stderr: # github.com/hashicorp/vault
warning: unable to find runtime/cgo.a
/usr/local/Cellar/go/1.6.2/libexec/pkg/tool/darwin_amd64/link: running clang failed: exit status 1
ld: warning: ignoring file /var/folders/2f/t653xfss3x34b__mf65rtrtr9v8rmg/T/go-link-053488021/go.o, file was built for unsupported file format ( 0x7F 0x45 0x4C 0x46 0x02 0x01 0x01 0x02 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 ) which is not the architecture being linked (x86_64): /var/folders/2f/t653xfss3x34b__mf65rtrtr9v8rmg/T/go-link-053488021/go.o
Undefined symbols for architecture x86_64:
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)


--> openbsd/amd64 error: exit status 2
Stderr: # github.com/hashicorp/vault
warning: unable to find runtime/cgo.a
/usr/local/Cellar/go/1.6.2/libexec/pkg/tool/darwin_amd64/link: running clang failed: exit status 1
ld: unknown option: -nopie
clang: error: linker command failed with exit code 1 (use -v to see invocation)


--> linux/amd64 error: exit status 2
Stderr: # github.com/hashicorp/vault
warning: unable to find runtime/cgo.a
/usr/local/Cellar/go/1.6.2/libexec/pkg/tool/darwin_amd64/link: running clang failed: exit status 1
ld: warning: ignoring file /var/folders/2f/t653xfss3x34b__mf65rtrtr9v8rmg/T/go-link-957473372/go.o, file was built for unsupported file format ( 0x7F 0x45 0x4C 0x46 0x02 0x01 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 ) which is not the architecture being linked (x86_64): /var/folders/2f/t653xfss3x34b__mf65rtrtr9v8rmg/T/go-link-957473372/go.o
Undefined symbols for architecture x86_64:
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)


--> netbsd/386 error: exit status 2
Stderr: # github.com/hashicorp/vault
warning: unable to find runtime/cgo.a
/usr/local/Cellar/go/1.6.2/libexec/pkg/tool/darwin_amd64/link: running clang failed: exit status 1
ld: warning: ignoring file /var/folders/2f/t653xfss3x34b__mf65rtrtr9v8rmg/T/go-link-841624487/go.o, file was built for unsupported file format ( 0x7F 0x45 0x4C 0x46 0x01 0x01 0x01 0x02 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 ) which is not the architecture being linked (i386): /var/folders/2f/t653xfss3x34b__mf65rtrtr9v8rmg/T/go-link-841624487/go.o
Undefined symbols for architecture i386:
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture i386
clang: error: linker command failed with exit code 1 (use -v to see invocation)


make: *** [bin] Error 1
```

**Comment by jefferai** [comment]
Huh. Which didn't work for some other people where the env var did.

All platforms except Darwin are built statically, and Darwin isn't only because Go itself can't build statically for Darwin. Adding those linker flags to all builds likely forces dynamic linking which indeed is problematic. So I will look into ways to add those for just Darwin. Does the binary work well still without the Web sense client installed?

**Comment by bkrodgers** [comment]
As far as I can tell, it still works fine without the websense client installed.  I was able to auth and do a list and read operation.  Anything in particular I should test that might be at risk of breaking with the change to external linking?

**Comment by jefferai** [comment]
I've been looking at https://golang.org/src/cmd/cgo/doc.go#L789

It basically has some clarification as to why `linkmode=external` helps but doesn't provide any guidance as to why this might be a good/bad idea to have on by default.

I'm going to ask the `gox` author about having per-OS/ARCH build flags. Right now there isn't a great way to accommodate this in the cross-compile build script.

**Comment by jefferai** [comment]
Hi @bkrodgers ,

I talked to the `gox` author and he'll try to find time soon to make per-arch/per-os flags, so we can turn on `linkmode=external` in `gox` for just macOS. Hopefully this should land before 0.6.1 so we can build it that way for that release.

**Comment by bkrodgers** [comment]
Great news, Jeff!  BTW, I also noticed that gox itself is affected by this same issue.  It's actually the reason my build of Vault was failing.  I thought the go toolchain itself was, but it actually looks like go will run, but it was failing when the build called gox.  If you've already got a related ticket open with gox, do you want to mention this, or should I open a new one?

**Comment by jefferai** [comment]
Hi Brian,

gox is used by Vault to cross compile. I didn't see anything indicating that gox was failing, rather that the build that gox kicked off using the Go toolchain was failing, because setting external linkmode when cross compiling is problematic. Did I miss something?

**Comment by bkrodgers** [comment]
That was failing too, but I was referring to trying to build Vault with the Websense client installed.  No matter what the linkmode is set to, I can't build Vault (or anything else using gox) because gox itself has this same issue where it throws bsdthread_register immediately on starting, when the Websense client is enabled.  I'm just saying that if they're going to add this ability to the next version of gox, it'd be nice if they also could build gox with it.  That may enable building Vault with the client installed as well.  I'm not sure if there are other issues with the build process hiding behind this one, but as it is now I have to disable Websense to build Vault.

**Comment by jefferai** [comment]
`gox` doesn't do releases -- you're supposed to install it with `go get github.com/mitchellh/gox`. AFAIK when `gox` is running it is merely toggling build flags and the os/arch env vars for each configured target. So I think that if you wanted to have `gox` not have this problem you'd need to locally fix how it's built by modifying the `go get` command with build flags -- see `go help build`.

I can see a problem happening if you got `gox` prebuilt via Homebrew or some other package manager, but then that's an issue to raise with the packager, not with `gox`.

**Comment by jefferai** [comment]
Migrating this to 0.6.2 milestone as there isn't any traction on the `gox` side so far.

**Comment by jefferai** [comment]
Good news -- it landed in `gox`. When the RC comes out can you test it and see if the build in the RC works for you?

**Comment by jefferai** [comment]
Modification to the above -- I can't build it for you since it requires an OSX machine in order to invoke the external linker. Is it an issue for you to build it yourself, now that doing so is pretty easy:
```
export GOX_DARWIN_AMD64_LDFLAGS="-linkmode=external -X github.com/hashicorp/vault/version.GitCommit='${GIT_COMMIT}${GIT_DIRTY}'"
```

**Comment by bkrodgers** [comment]
Sure, I can do it. Is that something I should be able to do now with what's in master?

So is that the main downside to the external linker vs internal with Go apps?  You lose the cross compile ability?

**Comment by bkrodgers** [comment]
OK, the build worked with that export enabled!  The vault executable it built appears to run fine with Websense installed.

BTW, the issue I mentioned with gox itself not running with Websense installed was resolved by installing it with this command:

`go get -ldflags '-linkmode=external' github.com/mitchellh/gox`

Any objection to adding a condition in the Makefile's bootstrap section to add that when building on Macs?  I'm happy to send in a PR for that.

**Comment by jefferai** [comment]
No objection!

**Comment by jefferai** [comment]
And yes -- I lose cross-compilation. Even when building an a Mac, which I don't currently have access to, we normally build inside a Docker container for various reasons, so that isn't possible with this change.

**Comment by bkrodgers** [comment]
Interesting.  Do you guys have a Mac available that you can start using for production releases, so that the binary distributions you guys put out are Websense friendly?  Or will this change ultimately just be a way for those of us subjected to Websense to solve it by building from source?

**Comment by jefferai** [comment]
I'll eventually have a Mac, and I could get my hands on one now if I needed to, but there are other issues. We'd like to continue cross-compiling in Docker, and I'm not sure we want to use external linkmode by default. Subtle changes in behavior can make hard-to-track-down bugs, especially if those changes happen deep in the Go runtime. At the same time, I don't think we want to confuse anyone by potentially having two builds for OSX available.

I'll discuss internally and see what we decide to do. For now, there's at least a way to build it from source.

**Comment by jefferai** [comment]
We've decided not to do this for now. Changing how the software is built introduces uncertainty in behavior that we don't understand and don't have the resources to vet. Given that Vault can be built now with the current changes, you can easily build supportable build internally.

I'm going to close the issue at this point but do reply if you need further help!

**Comment by lqyevil** [comment]
DLP-MacOS-v8.2.2525-0817  has solved this problem.

**Comment by jefferai** [comment]
Oh neat!
